### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by maxcorbeau** [refinedev/refine#4848]
Hi guys,

_**First of all thanks for refine**, I love your approach to a flexible CRUD framework, reason I could never adopt [react-admin](https://github.com/marmelab/react-admin) as it was restrictive in so many ways. Within a few minutes I was able to create a [multi-tenancy](https://refine.dev/docs/advanced-tutorials/multi-tenancy/appwrite/) CRUD setup with custom routing, which react-admin wouldn't let me. I can also implement custom functionality via URL while cohabiting with refine, which once again react-admin [wouldn't let me](https://github.com/marmelab/react-admin/issues/6923)._

**I am opening this ticket as I am experiencing performance issues**:
 - if I run [lighthouse](https://developer.chrome.com/docs/lighthouse/overview/) on the official [Finefoods MUI demo](https://example.mui.admin.refine.dev/) I get **~25-30/100 performance score**
 - I get this either on the **official hosted** version or a **local instance**
 - When interacting with the demo, **everything feels very sluggish**
 - For instance, on the `orders` page, when interacting with the `status` filter, there is a noticable several 100ms delay when first clicking the dropdown and also when selecting a value, whilst the filter itself is not live (you have to press the "FILTER" CTA to apply it)
 - I run latest Chrome on a powerful PC (no other app/website feels slow) so I don't think my machine is to blame
 - In fact, [pagespeed.web.dev](https://pagespeed.web.dev/analysis/https-example-mui-admin-refine-dev-orders/8kerm4esjx?form_factor=mobile) confirms similar poor performance on the login page alone

![refine_poor_performance_local](https://github.com/refinedev/refine/assets/24207660/0e91935a-fab9-4084-970c-aa886fb2ecb8)
![refine_poor_performance_hosted](https://github.com/refinedev/refine/assets/24207660/ccb4a63e-32e9-4a71-a7a0-dcea5a4b29bf)

So naturally my 2 questions are:  

**Why such performance issues? What can be done to address that?**

**Comment by BatuhanW** [comment]
Hello @maxcorbeau thanks for the detailed issue, initially, we believe most of the performance overhead coming from the UI library, however it needs a through investigation, we'll do it and get back to you.

**Comment by salihozdemir** [comment]
Hey @maxcorbeau ,

Thanks for the kind words. We're really thrilled that it's been useful for you!

I have investigated the performance issues as you mentioned. Before I share my findings, I want to clarify that these performance problems aren't directly connected to refine itself.

To start, I was curious about the performance scores of the examples built with Material UI. So, I checked out the official Material UI [templates](https://mui.com/templates/). I chose a few of them randomly and tested their performance scores. Here's what I discovered:

| Link                                                    | Score |
| ------------------------------------------------------- | ----- |
| https://minimals.cc/post                                | 27    |
| https://material-kit-pro-react.devias.io/dashboard/blog | 41    |
| https://thefront.maccarianagency.com/blog-reach-view    | 61    |
| https://mantisdashboard.io/login                        | 57    |

Interestingly, I expected better scores for the official Material UI templates. However, even these templates, which are essentially static pages without complex logic or data fetching libraries, didn't perform all that well.

Then, I wanted to figure out what sets refine projects apart from standard Material UI projects. The only difference I found is that refine projects include the core package (`@refinedev/core`), which is around 200KB when minified and approximately 57KB when minified and gzipped. Additionally, there's the `@refinedev/mui` UI package, sized at around 250KB when minified and 69KB when minified and gzipped.

In reality, these bundle sizes shouldn't be causing performance issues. Just to be absolutely sure, I examined the `base-material-ui` example. On average, this example achieved a performance score of around 80+. This tells us that the bundle size isn't likely the root cause of the problem.

Then, I've tried to increase the performance score of the Finefoods MUI demo. I've refactored the project using [React.lazy](https://reactjs.org/docs/code-splitting.html#reactlazy) and [React.Suspense](https://react.dev/reference/react/Suspense) APIs to lazy load the components. This approach involves breaking down the codebase into smaller, manageable chunks to optimize performance. Only with this, the performance score increased with following results:

| Page                                      | Old Score | New Score |
| ----------------------------------------- | --------- | --------- |
| https://finefoods-mui.refine.dev/         | 45        | 60        |
| https://finefoods-mui.refine.dev/orders   | 43        | 52        |
| https://finefoods-mui.refine.dev/users    | 51        | 70        |
| https://finefoods-mui.refine.dev/products | 48        | 70        |
| https://finefoods-mui.refine.dev/login    | 64        | 71        |

As you can see, the performance score increased significantly. However, I applied just one of the many possible optimizations. I'm sure that there are many other ways to improve the performance score even further. For example, using CDN for the images etc.

To sum up, while develop the demo examples, we focused on the showcase of the refine features. We didn't pay much attention to the performance score. So, you can improve the performance score of your project to be considerably better than the examples.

I hope this helps to clarify the situation a bit :pray:

**Comment by maxcorbeau** [comment]
Hey @salihozdemir, thanks so much for this thorough investigation. Indeed surprising to see the official MUI demos perform so poorly. This reassures me that the perf issues come indeed from the UI libraries and not refine itself.

**Would you be able to share a small example on how you refactored the project to use `React.lazy` and `React.suspense`?**

Many thanks

**Comment by BatuhanW** [comment]
Hey @maxcorbeau my friend @salihozdemir will get back to you, but in the meantime, I thought you might find this article useful:

https://dev.to/franklin030601/code-splitting-in-react-js-4o2g
